### PR TITLE
Model stdlib's IO package.

### PR DESCRIPTION
**Pull request by porcupineyhairs**
This PR models the golang stdlib io package. It depends on #129.

Here's a brief of changes made. 
1. I have moved `Reader.Read`,`Writer.Write`, `WriteString` which Max wrote in #129 into a new `IO.qll` file. I have also added a few doc comments on them.
1. I have further modelled multiple function definitions found in the `io` package. From what I can see, this completes the entire package.
1. I am yet to add tests but I would like a review before that.

Also, please note that I have based my PR on #129. Since, the PR is not yet merged, you may have to cherry-pick the commits before merging.

### CODE REVIEW

File: ql/src/semmle/go/frameworks/IO.qll [hunk 1]
```diff
@@ -0,0 +1,238 @@
+import go
+
+/** Provides models of commonly used functions in the `io` package. */
+module IO {
+  // A Taint Model for stdlib io Copy functions.
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:5]: This class and all the ones below are private, and the comment doesn't really add anything, so I'd suggest removing it here and for the other classes. (And as a reminder, `/** Qldoc comments look like this. */`.)

File: ql/src/semmle/go/frameworks/IO.qll [hunk 2]
```diff
@@ -0,0 +1,238 @@
+import go
+
+/** Provides models of commonly used functions in the `io` package. */
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:3]: Could you please either duplicate this comment at the top of the file to make sure the (implicit) file module is documented, or move this whole module back into `Stdlib.qll`? (I'd prefer the latter.)

File: ql/src/semmle/go/frameworks/IO.qll [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+module IO {
+  // A Taint Model for stdlib io Copy functions.
+  private class Copy extends TaintTracking::FunctionModel, Function {
+    // Include the Copy and CopyBuffer functions but exclude the
+    // CopyN function as it can't be realistically determined how long
+    // a buffer is enough for a particular exploit.
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:9]: That's an interesting point, but I'd like to see a false positive first before excluding it.

File: ql/src/semmle/go/frameworks/IO.qll [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    Copy() {
+      // func Copy(dst Writer, src Reader) (written int64, err error)
+      // func CopyBuffer(dst Writer, src Reader, buf []byte) (written int64, err error)
+      hasQualifiedName("io", "Copy") or
+      hasQualifiedName("io", "CopyBuffer")
+      // ignored
+      // func CopyN(dst Writer, src Reader, n int64) (written int64, err error)
+    }
+
+    // If the reader is tainted, the writer is considered tainted.
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:19]: I don't think we need this comment and similar ones below; they don't add anything beyond the code.

File: ql/src/semmle/go/frameworks/IO.qll [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+  // A Taint Model for the stdlib io Pipe functions.
+  private class Pipe extends TaintTracking::FunctionModel, Function {
+    Pipe() {
+      // func Pipe() (*PipeReader, *PipeWriter)
+      hasQualifiedName("io", "Pipe")
+    }
+
+    // If a pipe reads from a tainted pipereader, the pipe considered tainted.
+    override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+      input.isParameter(0) and output.isParameter(1)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:34]: Should this be `input.isResult(0) and output.isResult(1)`?

File: ql/src/semmle/go/frameworks/IO.qll [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    }
+  }
+
+  // A Taint Model for the stdlib io WriteString function.
+  private class WriteString extends TaintTracking::FunctionModel {
+    WriteString() { this.hasQualifiedName("io", "WriteString") }
+
+    override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+      // func WriteString(w Writer, s string) (n int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:58]: I think I'd prefer to put this comment into the characteristic predicate like for the other classes.

Comment (max-schaefer) [reply ql/src/semmle/go/frameworks/IO.qll:58]: Same below.

File: ql/src/semmle/go/frameworks/IO.qll [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+      input.isParameter(1) and output.isParameter(0)
+    }
+  }
+
+  // A Taint Model for the stdlib io ByteReader interface.
+  private class ByteReaderReadByte extends TaintTracking::FunctionModel, Method {
+    ByteReaderReadByte() { this.implements("io", "ByteReader", "ReadByte") }
+
+    // If the byteReader is tainted, consider the returned bytes tainted.
+    override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp) {
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:68]: The parameter names seem to be `input` and `output` above; perhaps adjust for consistency?

File: ql/src/semmle/go/frameworks/IO.qll [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+      // ReadByte() (byte, error)
+      inp.isReceiver() and outp.isParameter(0)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:70]:
```suggestion
      inp.isReceiver() and outp.isResult(0)
```

File: ql/src/semmle/go/frameworks/IO.qll [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+      // ReadFrom(r Reader) (n int64, err error)
+      // Note: The n here is the number bytes read and not the actual data.class
+      // this is inkoved in a manner similar to that shown below.
+      // var w io.Writer
+      // if w, ok := w.(io.ReaderFrom); ok {
+      //    m, err := w.ReadFrom(b.rd)
+      //    n += m
+      //    return n, err
+      // }
+      // Hence, in this case, it is the receiver which is tainted and not the result.
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:158]: Is this comment really needed? I think it's fairly clear from the type signature that this is what's happening.

File: ql/src/semmle/go/frameworks/IO.qll [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  // A Taint Model for the stdlib io RuneReader interface
+  private class RuneReaderReadRune extends TaintTracking::FunctionModel, Method {
+    RuneReaderReadRune() { this.implements("io", "RuneReader", "ReadRune") }
+
+    // If the RuneReader is tainted, consider the returned runes as tainted
+    override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp) {
+      // ReadRune() (r rune, size int, err error)
+      inp.isReceiver() and outp.isParameter(0)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:170]:
```suggestion
      inp.isReceiver() and outp.isResult(0)
```

File: ql/src/semmle/go/frameworks/IO.qll [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+    // If the string is tainted, consider the writer tainted
+    override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp) {
+      // WriteString(s string) (n int, err error)
+      inp.isParameter(0) and outp.isReceiver()
+    }
+  }
+
+  private class WriterWrite extends TaintTracking::FunctionModel, Method {
+    WriterWrite() { this.implements("io", "Writer", "Write") }
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:197]: Perhaps add a comment with the type signature? I know it's my code and I didn't include it, but I'm finding it helpful for the other classes.

File: ql/src/semmle/go/frameworks/IO.qll [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+  // A Taint Model for the stdlib io MultiWriter function
+  private class MultiWriter extends TaintTracking::FunctionModel, Function {
+    MultiWriter() {
+      // func MultiWriter(writers ...Writer) Writer
+      hasQualifiedName("io", "MultiWriter")
+    }
+
+    // If the primary writer is tainted, all otehr writers are tainted.
+    override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+      input.isParameter(0) and output.isParameter(any(int i | i > 0))
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/IO.qll:213]: I don't understand this flow specification. Shouldn't it be `input.isResult() and output.isParameter(_)`, i.e., if the multi-writer is tainted then all of the constituent writers are tainted as well?

File: ql/src/semmle/go/frameworks/IO.qll (hunk 3 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:9]: For a normal webapp, you can imagine a case where an untrusted parameter or something flows into a XSS sink through a copyN call. Now, there for most purposes, if the n is small say 5-10 bytes, regardless of whether the input is sanitised by other means, the flow does not lead to any security issues as an attacker can't frame a decent XSS Payload with just 5 bytes.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 12 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:213]: Corrected.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 11 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:197]: done.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 10 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:170]: done.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 5 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:34]: changed.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 6 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:58]: changed.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 7 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:68]: done!

File: ql/src/semmle/go/frameworks/IO.qll (hunk 8 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:70]: done!

File: ql/src/semmle/go/frameworks/IO.qll (hunk 9 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:158]: removed.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 2 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:3]: I created a separate file as `Stdlib.qll` was becoming big. Ayways, I have moved this back there.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 4 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:19]: removed

File: ql/src/semmle/go/frameworks/IO.qll (hunk 5 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll:34]: my bad. changed.

File: ql/src/semmle/go/frameworks/IO.qll (hunk 3 above)
Comment (porcupineyhairs) [reply ql/src/semmle/go/frameworks/IO.qll]: On second thought, I have included this with the PR. I think the True positive/false positive ratio would favor this change.

**Comment by porcupineyhairs** [comment]
I have added the test now. I am also modelling functions from the bufio and other pacakges for which I will send in another PR soon.

**Comment by sauyon** [comment]
Can you rebase or merge this over master now that the #129 has been

**Comment by porcupineyhairs** [comment]
Rebased and squashed for easier merging.

**Comment by max-schaefer** [comment]
I've started an evaluation, once that is done I think this is ready to merge.

**Pull request merged**